keps/sig-scheduling/3990-pod-topology-spread-fallback-mode/README.md

```diff
@@ -0,0 +1,1023 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+Your cluster has the cluster autoscaler 
+and you widely use Pod Topology Spread with `WhenUnsatisfiable: DoNotSchedule` for zone to strongthen workloads against the zone failure.
+And if the cluster autoscaler fails to create new Node for Pods due to the instance stockout, 
+you want to fallback from DoNotSchedule to ScheduleAnyway 
+because otherwise you'd hurt the availability of workload to achieve a better availability via Pod Topology Spread.
+That's putting the cart before the horse.
+
+In this case, you can use `ScaleUpFailed` in `fallbackCriteria`,
+to fallback from DoNotSchedule to ScheduleAnyway 
+
+```yaml
+topologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    fallbackCriteria:
+      - ScaleUpFailed
+    labelSelector:
+      matchLabels:
+        foo: bar
+```
+
+#### Story 2
+
+Your cluster doesn't have the cluster autoscaler 
+and has some low-priority Pods to make space (often called overprovisional Pods, balloon Pods, etc).
+Basically, you want to leverage preemption to achieve the best distribution as much as possible, 
+so you have to schedule Pods with `WhenUnsatisfiable: DoNotSchedule`.
+But, you don't want to make Pods unschedulable by Pod Topology Spread if the preemption won't make Pods schedulable.
+
+```yaml
+topologySpreadConstraints:
+  - maxSkew: 1
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    fallbackCriteria:
+      - PreemptionFailed
+    labelSelector:
+      matchLabels:
+        foo: bar
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### the fallback could be done when it's actually not needed.
+
+Even if the Pod is rejected by plugins other than Pod Topology Spread,
+when one of specified criteria is satisfied, the scheduler fallbacks from DoNotSchedule to ScheduleAnyway.
```

Comment:
For example, let's say PodA has a required Pod Topology Spread. When PodA's first scheduling, all Nodes went through Pod Topology Spread filter plugin, but they were rejected by other plugins (e.g., the resource fit filter plugin because of insufficient resources.), and PodA got unschedulable state as a result. Then, the cluster autoscaler tried to create a new Node for PodA but it couldn't make it because of stockout.

In this case, Pod Topology Spread doesn't need to fallback actually because it said OK to all Nodes in the previous scheduling cycle -- meaning Pod Topology Spread isn't the cause of unschedulable in the past scheduling cycle and the fallback won't make PodA schedulable. 
But, the problem is that Pod Topology Spread cannot see why this Pod was rejected in the previous scheduling cycle. It only understands that PodA was rejected by _someone_ in the previous scheduling, and the cluster autoscaler couldn't create a new Node.